Support for grpc bindings generator for Java; refreshed C++ generator to match grpc APIs in v 1.8

yfinkelstein opened the pull request:
I left a detailed message here: grpc/grpc-java#2510 

Thank you for submitting a PR!

Please make sure you include the names of the affected language(s) in your PR title.
This helps us get the correct maintainers to look at your issue.

If you make changes to any of the code generators, be sure to run
`cd tests && sh generate_code.sh` (or equivalent .bat) and include the generated
code changes in the PR. This allows us to better see the effect of the PR.

If your PR includes C++ code, please adhere to the Google C++ Style Guide,
and don't forget we try to support older compilers (e.g. VS2010, GCC 4.6.3),
so only some C++11 support is available.

Include other details as appropriate.

Thanks!

yfinkelstein commented:
I'm unclear on why CLA test is not passing. The email address on by github account matches by gmail account with which I signed CLA that *was* found per above, and my additional email on google account matches secondary on github account. At this point there is nothing else I can think of doing and need somebody at google to take a look at.

yfinkelstein commented:
I signed it!

aardappel commented on CMakeLists.txt:
```diff
@@ -151,7 +153,7 @@ elseif(MSVC)
   # warning C4512: assignment operator could not be generated
   # warning C4316: object allocated on the heap may not be aligned
-  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /W4 /WX /wd4512 /wd4316")
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /W4 /WX /wd4512 /wd4316 /wd4722")
```
please document above what this warning is.

aardappel commented on appveyor.yml:
```diff
@@ -12,7 +12,7 @@ environment:
 
   matrix:
-    - CMAKE_VS_VERSION: "10 2010"
+    #- CMAKE_VS_VERSION: "10 2010"
```
can't do this.

aardappel commented on src/idl_gen_grpc.cpp:
```diff
@@ -23,311 +23,357 @@
-        break;
-      }
+    
+    class FlatBufMethod : public grpc_generator::Method {
```
Why was this indented and reformatted? Please undo.

yfinkelstein commented:
I resolved the issues that you flagged. Here is the summary:

1. Restored compilation with msvc 2010. This required additional syntax fixing in both `c++` and `java` grpc generators. Both of these files came from grpc and grpc-java projects respectively. Moving forward, if there are additional changes the GPRC APIs somebody wold have to follow my path and merge original code with the modified versions here.

2. Added a whole new top level folder called java-grpc. It's a multi-module maven project which simplifies maintenance of 3 separate java maven artifacts: flatbuffers-java, flatbuffer-java-grpc, and flatbuffers-compiler. The compiler project publishes flatc binaries to maven in tar,gz form (zip on windows). There is a separate archive for each of the 3 platforms: Linux, osx, and Windows. The 2 java modules a test-dependent on flatbuffers-compiler and if the test fail the publishing would fail too. See https://github.com/yfinkelstein/flatbuffers/blob/master/java-grpc/flatbuffers-java-grpc/pom.xml#L57 for how the compiler is used at test time and https://github.com/yfinkelstein/flatbuffers/blob/master/java-grpc/flatbuffers-java-grpc/src/test/java/example/MonsterGrpcTest.java#L78 for an example of GRPC client and service based on FB. 

Note that a duplicated com.google.flatbuffers.*.java files under java-grpc for now. I did that because I believe flatbuffers-java needs to move there and removed from top-level java folder. Also, manual Java tests should be removed as the proper way to test Java code is inside of maven build which is what I did. However, I'm not sure how you will take this. If you disagree, just remove java-grpc/flatbuffers-java folder and remove this line: https://github.com/yfinkelstein/flatbuffers/blob/master/java-grpc/pom.xml#L203


Everything is working, however the real challenge is ahead: you need to find the place to stick maven publishing of the new artifact into the release process somewhere. Basically, all you need to do is to invoke `mvn deploy` inside of grpc-java folder. However, the problem is that in order to publish flatc binary for the 3 platforms you need to `have` the binaries stored locally at the time of maven deploy. How that can be accomplished - I have no idea. For now, I made the assumption that flatc binary that needs to be published is located two levels higher than java-grpc/flatbuffers-compiler folder i.e. at the root folder of the project. That assumption and related considerations is documented here: https://github.com/yfinkelstein/flatbuffers/blob/master/java-grpc/flatbuffers-compiler/pom.xml#L20

I spent quite a bit of time on this PR and would love to see it go through. Later we can fix any build-related nuances - almost certainly we may need to add a bit of automation to publish maven artifact s every time you have an official release.

Just to emphasize: existence of current flatc binary in maven is *very big deal* for adoption of flatbuffers in Java where developers are accustomed to having code generation at the time of the build rather than before the build. Doing this before the build requires them to install flatc on their build machines through some ad hoc mechanisms which is much harder and leads to various compatibility/versioning issues.

yfinkelstein commented:
One more thing. I added `.clang-format` and `.editorcondig` at the top of the project as they were missing. Without these files I simply could not merge my changes and maintain original formatting. I copied these files from gprc so they should follow google style. But do make sure my assumptions about tab length etc are correct.  `src/idl_gen_grpc.cpp` was reformatted by clang-format and hence quite a few lines have changed - I simply have no other way.

aardappel commented on java-grpc/flatbuffers-compiler/pom.xml:
```diff
@@ -0,0 +1,49 @@
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
```
Why can't this be under `grpc/` like everything else? Also, it seems like this is conflating 2 issues: making `flatc` available for Java users, and support for gRPC in Java, ideally those would be independent, and independent PRs. This makes it harder to review. It is hard for me to see if using `flatc` in Java without gRPC is possible, for example.

aardappel commented on java-grpc/flatbuffers-compiler/src/assembly/windows.xml:
```diff
@@ -0,0 +1,17 @@
+    <fileSet>
+      <directory>${basedir}/../..</directory>
+      <includes>
+        <include>flatc.exe</include>
```
If this means it is going to require me to build these on 3 platforms, then pull them back to Linux to be able to run Maven, then that is very inconvenient, and not sure that is something I want to commit to.

aardappel commented on java-grpc/flatbuffers-java-grpc/src/test/flatbuffers/monster-service.fbs:
```diff
@@ -0,0 +1,41 @@
+// Example IDL file for our monster's schema.
```
Can this be in `grpc/tests` like the other languages so far, and use the existing schema, again like the other languages?

aardappel commented on java-grpc/flatbuffers-java/src/main/java/com/google/flatbuffers/Constants.java:
```diff
@@ -0,0 +1,42 @@
+/*
```
Sorry, we cannot have this kind of duplication, they'll have to be pulled from their original location.

yfinkelstein replied:
My intent was to consolidate all Java client codec in a single directory and build and test with a single maven project. In that case you would delete the existing Java folder and there would be no duplication.
If you don’t like it, where do you suggest to put pom.xml of flatbuufers-java-grpc? Also, once you have 2 or more jars where one depends on another you really need a parent pom file to keep things consistent. That’s what I achieved with my multi-module maven project.
As for your comment about publishing flatc for multiple platforms - that is the hardest problem and I expected the difficulties. But we really need to figure this out. Otherwise we are back in the woods with developer needing to get flatc on his own

aardappel replied:
It be preferable to keep things in existing locations, and keep things testable without Maven. I do not know what the requirements are for pom files.. most build systems I know can address files from anywhere, is that not the case in Maven?

yfinkelstein commented:
Ok, I will be submitting multiple smaller PRs:
1. For gprc/cpp_generator.cc to become compatible with grpc 1.7/1.8
2. Small bug fix to idl_gen_json_schema.cpp 
3. java grpc code generator (flatc cpp code changes)
4. flatbuffers-java-grpc maven jar
5. flatbuffers-compiler for deploying fratc binaries to maven
6. .clang-formatter, etc

Question: items 3 and 4 above are sufficiently different in nature (c++ vs Java) however to test item 4 item 3 needs to be in place. Do you want to keep them in separate PRs or combine? Personally, I'd prefer to keep them separate since there may be more questions about item test and I want to get the flatc changes in and resolve them.

I will keep this PR open simply to keep the context open but once the smaller PRs go in I will close this PR

aardappel commented:
Thanks for the multiple PRs, that will make it much quicker to review and merge. Start with the less contentious ones :)

We can keep 3 & 4 separate yes.

yfinkelstein commented:
With regards to publishing flatc compiler binaries to maven and your comment

>If this means it is going to require me to build these on 3 platforms, then pull them back to Linux to be able to run Maven, then that is very inconvenient, and not sure that is something I want to commit to.

as a reference point consider that protobuf/protoc is doing exactly that:
http://repo1.maven.org/maven2/com/google/protobuf/protoc/3.5.0/

They must have found a sustainable method of building executables on multiple platforms and publishing them to maven central. I fond this page that provides a good hint:
https://github.com/google/protobuf/tree/master/protoc-artifacts

If I get it right, they simply cross compile using gcc:
* for windows they build with in Linux with MinGW-w64 (x86_32, x86_64) 
* for osx, it says nothing specific but I guess they run the build and publish from a VM with osx (https://github.com/geerlingguy/macos-virtualbox-vm)  Or perhaps google has a form of mac servers ...

aardappel commented:
Oh, I am sure it can be done, but setting that up is a bit of work. Sadly FlatBuffers doesn't have quite the headcount that Protobuf does :)

aardappel commented:
I'm closing this to indicating it will not be merged, in favor of individual PRs. Feel free to keep the discussion going though :)

yfinkelstein commented:
So, now that all my smaller PRs are merged I'd like to revisit the only topic left: publishing flatc binaries to maven. Per earlier discussion, protobuf is doing this and they have detailed info/scripts on how they are doing it. The only part that's not documented is how they are spinning the VMs with osx and windows.

yfinkelstein commented:
I was looking closer the protobuf maven plugin code and discovered that they also support generation of code for all languages supported by protobuf and beyond: they support custom plugins written in Java. The protobuf maveb plugin can wrap the Java protobuf plugin with a native binary that protobuf complier can pick up if it's placed in the write directory and corresponding command line args are supplied. Furthermore, to simply maven distribution they bundle the plugin with a wrapper script `mvnw` https://github.com/xolstice/protobuf-maven-plugin/blob/master/mvnw which can fetch maven distro on the fly. 

What all this means is that this protobuf maven plugin effectively complements protoc to make it easy to download and use on all platforms  where Java is supported.  But this plugin does not come with protoc binary itself and instead relies on maven repo to download it from: http://repo1.maven.org/maven2/com/google/protobuf/protoc/3.5.0/

I'm convinced that a modest-sized one time effort to set up a automatic release-time flatc publishing to maven for 3 popular platforms will pay off in a form of wider adoption of flatbuffers.

aardappel commented:
@yfinkelstein I certainly agree this would be great to have, but I don't think I have the bandwidth to create and maintain such services here internally.

If anything, I wonder if it wouldn't be easier to use the existing Travis and AppVeyor CI that we have to generate build artifacts for `flatc` on every run, than a script used by Maven can pull from.

I don't see the point in using Maven for other languages, as they tend to have their own build/deploy systems. Using a foreign such system in addition to your own is usually unwelcome.
